x/reward/types/reward.go

```diff
@@ -0,0 +1,595 @@
+package types
+
+import (
+	"errors"
+	fmt "fmt"
+	"reflect"
+	"strings"
+	time "time"
+
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	"github.com/provenance-io/provenance/internal/pioconfig"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// DefaultStartingRewardProgramID is 1
+const DefaultStartingRewardProgramID uint64 = 1
+
+// Constants pertaining to a RewardProgram object
+const (
+	MaxDescriptionLength int = 10000
+	MaxTitleLength       int = 140
+	DayInSeconds         int = 60 * 60 * 24
+)
+
+var (
+	_ RewardAction = &ActionDelegate{}
+	_ RewardAction = &ActionTransfer{}
+	_ RewardAction = &ActionVote{}
+)
+
+const (
+	ActionTypeDelegate = "ActionDelegate"
+	ActionTypeTransfer = "ActionTransfer"
+	ActionTypeVote     = "ActionVote"
+)
+
+// RewardAction defines the interface that actions need to implement
+type RewardAction interface {
+	ActionType() string
+	Evaluate(ctx sdk.Context, provider KeeperProvider, state RewardAccountState, event EvaluationResult) bool
+	PreEvaluate(ctx sdk.Context, provider KeeperProvider, state RewardAccountState) bool
+	PostEvaluate(ctx sdk.Context, provider KeeperProvider, state RewardAccountState) bool
+	GetBuilder() ActionBuilder
+}
+
+// ============ Shared structs ============
+
+type ABCIEvent struct {
+	Type       string
+	Attributes map[string][]byte
+}
+
+type EventCriteria struct {
+	Events map[string]ABCIEvent
+}
+
+// NewEventCriteria Performs a shallow copy of the map
+// We are assuming this takes ownership of events
+func NewEventCriteria(events []ABCIEvent) *EventCriteria {
+	criteria := EventCriteria{}
+	if len(events) == 0 {
+		return &criteria
+	}
+	criteria.Events = make(map[string]ABCIEvent)
+	for _, event := range events {
+		criteria.Events[event.Type] = event
+	}
+	return &criteria
+}
+
+func (ec *EventCriteria) MatchesEvent(eventType string) bool {
+	// If we have no Events then we match everything
+	if ec.Events == nil {
+		return true
+	}
+
+	// If we don't have the event then we don't match it
+	_, exists := ec.Events[eventType]
+	return exists
+}
+
+func (ec *ABCIEvent) MatchesAttribute(name string, value []byte) bool {
+	attribute, exists := ec.Attributes[name]
```

Comment:
Probably need to make sure `ec.Attributes` isn't nil first.

Reply:
If ec.Attributes is nil then exists will be false.

https://go.dev/play/p/K26sK9wUrlZ?v=goprev

Reply:
Oh! Nice! I could have sworn I've gotten nil pointer errors when trying to access a nil map before. Maybe I'm misremembering that or maybe they fixed it in 1.18. Either way. Nice.